tests/test_thumbs_function.py

```diff
@@ -0,0 +1,117 @@
+from pathlib import Path
+
+import boto3
+import pytest
+from moto import mock_aws
+
+# Constants used for testing
+TEST_BUCKET = "leaflet-test-thumbs"
+TEST_IMAGES_PATH = Path(__file__).parent / "test_images"
+
+
+@pytest.fixture
+def mock_s3_bucket(monkeypatch):
+    monkeypatch.setenv("LEAFLET_IMAGES_BUCKET_NAME", TEST_BUCKET)
+
+    with mock_aws():
+        s3 = boto3.client("s3")
+        s3.create_bucket(Bucket=TEST_BUCKET)
+        yield s3
+        # Clean up after
+        objects = s3.get_paginator("list_objects_v2")
+        objects_iterator = objects.paginate(Bucket=TEST_BUCKET)
+        for page in objects_iterator:
+            if "Contents" in page:
+                objects = [{"Key": obj["Key"]} for obj in page["Contents"]]
+                s3.delete_objects(
+                    Bucket=TEST_BUCKET, Delete={"Objects": objects}
+                )
+        s3.delete_bucket(Bucket=TEST_BUCKET)
+
+
+IMAGES_TO_TEST = (
+    TEST_IMAGES_PATH / "front_test.jpg",
+    TEST_IMAGES_PATH / "heif_test.heif",
+    TEST_IMAGES_PATH / "tiff_test.tiff",
+)
+
+
+@pytest.mark.parametrize("image_filename", IMAGES_TO_TEST)
+def test_handle_s3_creates_thumbnails(
+    image_filename,
+    mock_s3_bucket,
+):
+    mock_s3_bucket.put_object(
+        Bucket=TEST_BUCKET,
+        Key=f"leaflets/{image_filename.name}",
+        Body=image_filename.read_bytes(),
+    )
+
+    event = {
+        "Records": [
+            {
+                "s3": {
+                    "bucket": {"name": TEST_BUCKET},
+                    "object": {"key": f"leaflets/{image_filename.name}"},
+                }
+            }
+        ]
+    }
+    from thumbs.handler import handle_s3
+
+    handle_s3(event, context=None, local=False)
+
+    all_objects = mock_s3_bucket.list_objects_v2(Bucket=TEST_BUCKET)
+    keys = [obj["Key"] for obj in all_objects.get("Contents", [])]
+    expected_filename = image_filename.with_suffix(".png").name
+    expected_starts = [
+        f"thumbs/350/crop=top/leaflets/{expected_filename}",
+        f"thumbs/150/crop=noop/leaflets/{expected_filename}",
+        f"thumbs/1000/crop=noop/leaflets/{expected_filename}",
+        f"thumbs/600/crop=center/leaflets/{expected_filename}",
+        f"thumbs/350/leaflets/{expected_filename}",
+        f"thumbs/600/leaflets/{expected_filename}",
+    ]
+
+    for expected_key in expected_starts:
```

Comment:
There's a couple of tests here where we are looping over a thing and making assertions in the loop. When we do this, we should start off by asserting the expected length of the thing first, otherwise an empty iterable passes the test, which is probably not what we intended.

Reply:
I understand this principle, but can you confirm that you think it's needed in this exact case?

I'm asking because the list I'm iterating over is defined in the line above this loop. So what you're suggesting would end up looking like:

```py
x = [1,2,3]
assert len(x) == 3
for i in x:
    ...
```

?

Reply:
Yeah you're right. Having read over it again, it is not actually a useful suggestion in either of these cases.